File: app/src/main/java/com/ntech/fourtop/ui/login/LoginViewModel.java
package com.ntech.fourtop.ui.login;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.ntech.fourtop.data.AppRepository;
import com.ntech.fourtop.network.pojo.ResponseToken;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

public class LoginViewModel extends ViewModel {

    private AppRepository repository;
    private CompositeDisposable disposables;
    private MutableLiveData<String> throwable;
    private MutableLiveData<String> userToken;
    private MutableLiveData<String> errorMessage;

    public LoginViewModel() {
        repository = AppRepository.get();
        disposables = new CompositeDisposable();
        throwable = new MutableLiveData<>();
        userToken = new MutableLiveData<>();
        errorMessage = new MutableLiveData<>();
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        disposables.clear();
    }

    public LiveData<String> getThrowable() {
        return throwable;
    }

    public LiveData<String> getUserToken() {
        return userToken;
    }

    public LiveData<String> getErrorMessage() {
        return errorMessage;
    }

    public void loginRequest(String login, String password) {
        Timber.d("Send login request");
        Disposable disposable = repository.loginRequest(login, password)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::onSuccess, this::onError);
        disposables.add(disposable);
    }

    private void onSuccess(ResponseToken response) {
        int status = response.getStatus();
        Timber.d("Login request status %d", status);

        switch (status) {
            case 200:
                userToken.setValue(response.getData().getUserToken());
                break;
            case 500:
                errorMessage.setValue(response.getErrorMsg());
                break;
        }
    }

    private void onError(Throwable t) {
        Timber.d(t);
        throwable.setValue(t.toString());
    }
}

File: app/src/main/java/com/emikhalets/voteapp/view/login/LoginViewModel.kt
package com.emikhalets.voteapp.view.login

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.emikhalets.voteapp.data.AppRepository
import com.emikhalets.voteapp.data.AppRepository.Companion.get
import com.emikhalets.voteapp.network.pojo.ResponseToken
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import timber.log.Timber

class LoginViewModel : ViewModel() {
    private val repository: AppRepository?
    private val disposables: CompositeDisposable
    private val throwable: MutableLiveData<String?>
    private val userToken: MutableLiveData<String?>
    private val errorMessage: MutableLiveData<String?>
    override fun onCleared() {
        super.onCleared()
        disposables.clear()
    }

    fun getThrowable(): LiveData<String?> {
        return throwable
    }

    fun getUserToken(): LiveData<String?> {
        return userToken
    }

    fun getErrorMessage(): LiveData<String?> {
        return errorMessage
    }

    fun loginRequest(login: String?, password: String?) {
        Timber.d("Send login request")
        val disposable = repository!!.loginRequest(login, password)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response: ResponseToken -> this.onSuccess(response) }) { t: Throwable -> onError(t) }
        disposables.add(disposable)
    }

    private fun onSuccess(response: ResponseToken) {
        val status = response.status
        Timber.d("Login request status %d", status)
        when (status) {
            200 -> userToken.setValue(response.data!!.userToken)
            500 -> errorMessage.setValue(response.errorMsg)
        }
    }

    private fun onError(t: Throwable) {
        Timber.d(t)
        throwable.value = t.message
    }

    init {
        repository = get()
        disposables = CompositeDisposable()
        throwable = MutableLiveData()
        userToken = MutableLiveData()
        errorMessage = MutableLiveData()
    }
}